File: himsha-sdk-dart/lib/src/programs/token.dart
import 'dart:typed_data';
import '../pubkey.dart';
import '../transaction.dart';

Uint8List _u64Le(BigInt n) {
  final buf = ByteData(8)..setUint64(0, n.toInt(), Endian.little);
  return buf.buffer.asUint8List();
}

Uint8List _concat(List<Uint8List> arrays) {
  final total = arrays.fold(0, (s, a) => s + a.length);
  final out = Uint8List(total);
  int offset = 0;
  for (final a in arrays) {
    out.setRange(offset, offset + a.length, a);
    offset += a.length;
  }
  return out;
}

Uint8List _optKey(HimshaPublicKey? key) =>
    key != null ? _concat([Uint8List.fromList([1]), key.bytes]) : Uint8List.fromList([0]);

class TokenProgram {
  static HimshaInstruction initializeMint(
    HimshaPublicKey mint,
    HimshaPublicKey authority,
    int decimals, [
    HimshaPublicKey? freezeAuthority,
  ]) =>
      HimshaInstruction(
        programId: ProgramIds.token,
        accounts: [AccountMeta.writable(mint, false)],
        data: _concat([
          Uint8List.fromList([0, decimals]),
          authority.bytes,
          _optKey(freezeAuthority),
        ]),
      );

  static HimshaInstruction initializeAccount(
    HimshaPublicKey account,
    HimshaPublicKey mint,
    HimshaPublicKey owner,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.token,
        accounts: [
          AccountMeta.writable(account, false),
          AccountMeta.readonly(mint, false),
          AccountMeta.readonly(owner, false),
        ],
        data: Uint8List.fromList([1]),
      );

  static HimshaInstruction mintTo(
    HimshaPublicKey mint,
    HimshaPublicKey destination,
    HimshaPublicKey authority,
    BigInt amount,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.token,
        accounts: [
          AccountMeta.writable(mint, false),
          AccountMeta.writable(destination, false),
          AccountMeta.readonly(authority, true),
        ],
        data: _concat([Uint8List.fromList([2]), _u64Le(amount)]),
      );

  static HimshaInstruction transfer(
    HimshaPublicKey source,
    HimshaPublicKey destination,
    HimshaPublicKey owner,
    BigInt amount,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.token,
        accounts: [
          AccountMeta.writable(source, false),
          AccountMeta.writable(destination, false),
          AccountMeta.readonly(owner, true),
        ],
        data: _concat([Uint8List.fromList([3]), _u64Le(amount)]),
      );

  static HimshaInstruction burn(
    HimshaPublicKey account,
    HimshaPublicKey mint,
    HimshaPublicKey owner,
    BigInt amount,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.token,
        accounts: [
          AccountMeta.writable(account, false),
          AccountMeta.writable(mint, false),
          AccountMeta.readonly(owner, true),
        ],
        data: _concat([Uint8List.fromList([4]), _u64Le(amount)]),
      );

  static HimshaInstruction closeAccount(
    HimshaPublicKey account,
    HimshaPublicKey destination,
    HimshaPublicKey owner,
  ) =>
      HimshaInstruction(
        programId: ProgramIds.token,
        accounts: [
          AccountMeta.writable(account, false),
          AccountMeta.writable(destination, false),
          AccountMeta.readonly(owner, true),
        ],
        data: Uint8List.fromList([9]),
      );
}

File: himsha-sdk-python/himsha_sdk/programs/token.py
"""HIMSHA Token Program helpers."""

import struct
from ..pubkey import HimshaPublicKey, PROGRAM_IDS
from ..transaction import AccountMeta, HimshaInstruction


def _u64(n: int) -> bytes:
    return struct.pack("<Q", n)


def _opt_key(key: HimshaPublicKey | None) -> bytes:
    return (b"\x01" + key.to_bytes()) if key else b"\x00"


def initialize_mint(
    mint: HimshaPublicKey,
    authority: HimshaPublicKey,
    decimals: int,
    freeze_authority: HimshaPublicKey | None = None,
) -> HimshaInstruction:
    data = bytes([0, decimals]) + authority.to_bytes() + _opt_key(freeze_authority)
    return HimshaInstruction(
        program_id=PROGRAM_IDS.token,
        accounts=[AccountMeta.writable(mint, is_signer=False)],
        data=data,
    )


def initialize_account(
    account: HimshaPublicKey, mint: HimshaPublicKey, owner: HimshaPublicKey
) -> HimshaInstruction:
    return HimshaInstruction(
        program_id=PROGRAM_IDS.token,
        accounts=[
            AccountMeta.writable(account, is_signer=False),
            AccountMeta.readonly(mint, is_signer=False),
            AccountMeta.readonly(owner, is_signer=False),
        ],
        data=bytes([1]),
    )


def mint_to(
    mint: HimshaPublicKey,
    destination: HimshaPublicKey,
    authority: HimshaPublicKey,
    amount: int,
) -> HimshaInstruction:
    return HimshaInstruction(
        program_id=PROGRAM_IDS.token,
        accounts=[
            AccountMeta.writable(mint, is_signer=False),
            AccountMeta.writable(destination, is_signer=False),
            AccountMeta.readonly(authority, is_signer=True),
        ],
        data=bytes([2]) + _u64(amount),
    )


def transfer(
    source: HimshaPublicKey, destination: HimshaPublicKey, owner: HimshaPublicKey, amount: int
) -> HimshaInstruction:
    return HimshaInstruction(
        program_id=PROGRAM_IDS.token,
        accounts=[
            AccountMeta.writable(source, is_signer=False),
            AccountMeta.writable(destination, is_signer=False),
            AccountMeta.readonly(owner, is_signer=True),
        ],
        data=bytes([3]) + _u64(amount),
    )


def burn(
    account: HimshaPublicKey, mint: HimshaPublicKey, owner: HimshaPublicKey, amount: int
) -> HimshaInstruction:
    return HimshaInstruction(
        program_id=PROGRAM_IDS.token,
        accounts=[
            AccountMeta.writable(account, is_signer=False),
            AccountMeta.writable(mint, is_signer=False),
            AccountMeta.readonly(owner, is_signer=True),
        ],
        data=bytes([4]) + _u64(amount),
    )


def close_account(
    account: HimshaPublicKey, destination: HimshaPublicKey, owner: HimshaPublicKey
) -> HimshaInstruction:
    return HimshaInstruction(
        program_id=PROGRAM_IDS.token,
        accounts=[
            AccountMeta.writable(account, is_signer=False),
            AccountMeta.writable(destination, is_signer=False),
            AccountMeta.readonly(owner, is_signer=True),
        ],
        data=bytes([9]),
    )
